### PR TITLE
Add current implementation

### PR DESCRIPTION
**Pull request by hartwork**
The idea with the pull request is to ease review. So **review welcome**.

Conscious decisions made include:
- Unlike the Python blueprint, we don't need a serial mode here: Fully concurrent is intended.

Things known to not be complete include:
- There are no automated tests, yet.
- The version is purely informational, I still need to learn more about versioning Go modules.
- The `fmt.Sprintf` calls where logging will be merged with the logging layer to simplify, probably.
- There is no CI yet. It will run a build, `gofmt`, `goimports`, and tests.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+/go-wait-for-it
+/go.sum
```
Comment (metalmatze) [inline .gitignore:2]: Usually this gets checked in, as it allows for reproducible builds.

File: cli/parser.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+type Config struct {
+	Addresses []syntax.Address
+	Argv      []string
+	Quiet     bool
+	Timeout   time.Duration
+}
+
+func report(err error) {
+	fmt.Println("Error:", err)
+}
```
Comment (metalmatze) [inline cli/parser.go:25]: I wouldn't create this extra func, just inline it for these two occasions.

File: main.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+)
+
+type ConnectResult struct {
+	address  syntax.Address
+	duration time.Duration
+	success  bool
+}
+
+func waitForAddress(address syntax.Address) chan bool {
+	available := make(chan bool)
```
Comment (metalmatze) [inline main.go:27]: Make sure to `close(available)` too. Otherwise the channel stays around and (especially for long lived programs) will become a problem.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	go func() {
+		for {
+			c, err := net.Dial("tcp", address.String())
+			if err == nil {
```
Comment (metalmatze) [inline main.go:31]: Probably?
```suggestion
			if err != nil {
```

Comment (metalmatze) [reply main.go:31]: And the surface these errors by returning `(bool, error)` with from the functions and not a channel.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+				c.Close()
+				break
+			}
+			time.Sleep(500 * time.Millisecond)
+		}
+	}()
+	return available
+}
+
+func waitForAddressWithTimeout(address syntax.Address, timeout time.Duration, results chan ConnectResult) {
```
Comment (metalmatze) [inline main.go:42]: The timeout can actually be more nicely implemented with a context.Context that carries a timeout via `context.WithTimeout(context.Background(), time.Second)`.
This is totally fine to do in a separate PR as improvement :blush:

File: main.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	case <-waitForAddress(address):
+		duration = time.Now().Sub(before)
+		success = true
+	case <-deadline:
+	}
+
+	results <- ConnectResult{address, duration, success}
+}
+
+func waitForMultipleAddressesWithTimeout(addresses []syntax.Address, timeout time.Duration, log logging.Log) bool {
```
Comment (metalmatze) [inline main.go:63]: I'd rather return an error than a boolean here.

File: main.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	success := true
+	results := make(chan ConnectResult)
+
+	for _, address := range addresses {
+		log.Neutral(fmt.Sprintf("Trying to connect to %s...", address))
+		go waitForAddressWithTimeout(address, timeout, results)
+	}
+
+	for replies := 0; success && replies < len(addresses); replies++ {
```
Comment (metalmatze) [inline main.go:72]: `for range` should probably be possible and more safe to do.

File: main.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+		result := <-results
+		if result.success {
+			log.Success(fmt.Sprintf("Connected to %s after %s.", result.address, result.duration))
+		} else {
+			success = false
+			log.Error(fmt.Sprintf("Failed to connected to %s for %s.", result.address, result.duration))
+			log.Error("Aborting...")
+		}
```
Comment (metalmatze) [inline main.go:80]: Flip it around checking for a failure and then return early with an error.

File: main.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	log.Error(fmt.Sprintf("Error: %v", err))
+
+	if exitError, ok := err.(*exec.ExitError); ok {
+		return exitError.ExitCode()
+	}
+
+	return 127
+}
+
+func main() {
```
Comment (metalmatze) [inline main.go:108]: Usually `func main()` goes all the way to the top to make it easiest for people to find the entrypoint to the program.

Comment (metalmatze) [reply main.go:108]: Probably that way, cause you're used to Python :upside_down_face:

**Review by metalmatze — CHANGES_REQUESTED** [review]
Pretty sweet for your first Go code. I can definitely see the Python influences, but nothing wrong with that. :)

**Comment by metalmatze** [comment]
Another thing that came to my mind while looking at this: With so few files, it's totally fine to put all code into `package main`, meaning the root of the project. Generally speaking there's probably something wrong too, if you only have one file in a package. This will also make it easier to avoid cyclic dependencies in the future.

File: .gitignore (hunk 1 above)
Comment (hartwork) [reply .gitignore:2]: I noticed a pile of lines for non-dependency packages in that file; the file is 285 lines in total. While I may not have anything secret in there, it felt like a violation of privacy. Any ideas about how to trim that file down to the list of all direct and indirect dependencies, only?

File: cli/parser.go (hunk 2 above)
Comment (hartwork) [reply cli/parser.go:25]: I want to have the two places in sync at all times so I think I'll stick to D.R.Y. here: The added complexity seems small enough to me for that to be okay.

File: .gitignore (hunk 1 above)
Comment (metalmatze) [reply .gitignore:2]: It should be direct and indirect dependencies. It only contains those, nothing more. I don't see how that should violate your privacy honestly (I'm curious). Basically all Go projects have these.

File: main.go (hunk 3 above)
Comment (hartwork) [reply main.go:27]: I'd have to close it in `waitForAddressWithTimeout`. I think I get in trouble with the timeout then: If the timeout hits, I close the channel, the goroutine succeeds and writes to the channel: Wouldn't that fail then and maybe even crash the program?

I out that to the test and it was indeed a problem, at least in my scenario.
Here's my `demo.go `:

```go
package main

import "time"

func writeForever(channel chan int) {
    for {
        channel <- 1
    }
}

func main() {
    channel := make(chan int)
    go writeForever(channel)
    close(channel)
    time.Sleep(10 * time.Second)
}
```

When I run it I get:

```console
# go build . && "./$(basename "$PWD")"
panic: send on closed channel

goroutine 5 [running]:
main.writeForever(0xc000016120)
        /tmp/tmp.pWKklKvfrd/demo.go:7 +0x37
created by main.main
        /tmp/tmp.pWKklKvfrd/demo.go:13 +0x58
```

Any ideas how to not have that problem, in general or in the very case of `go-wait-for-it` here?

File: main.go (hunk 4 above)
Comment (hartwork) [reply main.go:31]: I re-checked: The condition is not flipped around. I want `waitForAddress` to try connecting to a specific TCP port until it succeeds to no error is the case where I need to close the socket and be done. Right now I don't need the error details and it's an internal function. I think I'll leave it this way.

File: main.go (hunk 7 above)
Comment (hartwork) [reply main.go:72]: I didn't think there was an easy version with `range` possible but I found one now, good idea. Will push in a minute.

File: main.go (hunk 8 above)
Comment (hartwork) [reply main.go:80]: Good idea, will push an update in a minute.

File: main.go (hunk 9 above)
Comment (hartwork) [reply main.go:110]: I think that's actually my roots in C: You could move it further up in Python no problem. I re-checked `A tour of Go` and all the examples with multiple functions I can see have `main` last. I think I'll stick with that for now.

File: main.go (hunk 6 above)
Comment (hartwork) [reply main.go]: Good idea, how do you like the substitute I just pushed?

**Comment by hartwork** [comment]
> Another thing that came to my mind while looking at this: With so few files, it's totally fine to put all code into `package main`, meaning the root of the project. Generally speaking there's probably something wrong too, if you only have one file in a package. This will also make it easier to avoid cyclic dependencies in the future.

Could you elaborate how merging things back into `main` could help reduce the risk of cyclic dependencies? I'm happy to help fight cyclic dependencies in general, I just don't get the idea how that change might help, yet. Thanks!

File: .gitignore (hunk 1 above)
Comment (hartwork) [reply .gitignore:2]: Okay, let's see If I can prove or disprove that these 285 contain unrelated packages. I have an idea, give me a few minutes.

Comment (hartwork) [reply .gitignore:2]: I tried it in fresh Debian container now and I get the same 285 lines for `go.sum` in there. Alright, I'll add the file.

File: main.go (hunk 5 above)
Comment (hartwork) [reply main.go:42]: Interesting! I'll go read https://blog.golang.org/context

Comment (hartwork) [reply main.go]: I grabbed some ideas from the post and other docs on contexts by now but I'm still too much in the dark with some key questions about it, to use it myself in code today. I think I'll delay the topic of contexts in Go until a moment with voice chat on the topic, in the future. Thanks for bringing it up, definitely of interest for any future HTTP server of mine in Go :smiley:
